metadata/src/main/java/org/apache/kafka/timeline/SnapshottableHashTable.java

```diff
@@ -0,0 +1,449 @@
+                 * If elements move from the top tier to the snapshot tier, then
+                 * we'll still find them and report them exactly once.
+                 *
+                 * Note that while I used 4 and 2 as example sizes here, the same pattern
+                 * holds for different powers of two.  The "snapshot slot" of an element
+                 * will be the top few bits of the top tier slot of that element.
+                 */
+                Iterator<Snapshot> iterator = snapshotRegistry.iterator(snapshot);
+                while (iterator.hasNext()) {
+                    Snapshot curSnapshot = iterator.next();
+                    HashTier<T> tier = curSnapshot.getDelta(SnapshottableHashTable.this);
+                    if (tier != null && tier.deltaTable != null) {
+                        BaseHashTable<T> deltaTable = tier.deltaTable;
+                        int shift = Integer.numberOfLeadingZeros(deltaTable.baseElements().length) -
+                            Integer.numberOfLeadingZeros(topTier.length);
+                        int tierSlot = slot >>> shift;
+                        BaseHashTable.unpackSlot(temp, deltaTable.baseElements(), tierSlot);
+                        for (T object : temp) {
+                            if (BaseHashTable.findSlot(object, topTier.length) == slot) {
+                                ready.add(object);
+                            }
+                        }
+                        temp.clear();
+                    }
+                }
+                slot++;
+            }
+            return true;
+        }
+
+        @Override
+        public T next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            return ready.remove(ready.size() - 1);
+        }
+    }
+
+    private final SnapshotRegistry snapshotRegistry;
+
+    SnapshottableHashTable(SnapshotRegistry snapshotRegistry, int expectedSize) {
+        super(expectedSize);
+        this.snapshotRegistry = snapshotRegistry;
+    }
+
+    int snapshottableSize(long epoch) {
+        if (epoch == LATEST_EPOCH) {
+            return baseSize();
+        } else {
+            Iterator<Snapshot> iterator = snapshotRegistry.iterator(epoch);
+            while (iterator.hasNext()) {
+                Snapshot snapshot = iterator.next();
+                HashTier<T> tier = snapshot.getDelta(SnapshottableHashTable.this);
+                if (tier != null) {
+                    return tier.size;
+                }
+            }
+            return baseSize();
+        }
+    }
+
+    T snapshottableGet(Object key, long epoch) {
+        T result = baseGet(key);
+        if (result != null && result.startEpoch() <= epoch) {
+            return result;
+        }
+        if (epoch == LATEST_EPOCH) {
+            return null;
+        }
+        Iterator<Snapshot> iterator = snapshotRegistry.iterator(epoch);
+        while (iterator.hasNext()) {
```

Comment:
It seems that we don't need to do the while loop. Perhaps we could just assert `hasNext() `here.

Reply:
The while loop is needed because we need to search forward in the snapshot list.

Basically this is an optimization for PUT.  When creating a new delta entry, rather than adding that delta entry to every previous snapshot, we only add it to the most recent snapshot.  Then, when reading from an older snapshot, we need to search forward from there to make sure we're seeing everything.